pkg/rtc/subscribedtrack.go

```diff
@@ -57,14 +62,28 @@ func (t *SubscribedTrack) SetPublisherMuted(muted bool) {
 	t.updateDownTrackMute()
 }
 
-func (t *SubscribedTrack) UpdateSubscriberSettings(enabled bool, quality livekit.VideoQuality) {
-	t.debouncer(func() {
-		t.subMuted.TrySet(!enabled)
-		t.updateDownTrackMute()
-		if enabled && t.dt.Kind() == webrtc.RTPCodecTypeVideo {
-			t.dt.SetMaxSpatialLayer(spatialLayerForQuality(quality))
-		}
-	})
+func (t *SubscribedTrack) UpdateSubscriberSettings(settings *livekit.UpdateTrackSettings) {
+	t.subMuted.TrySet(settings.Disabled)
+	t.settings.Store(settings)
+	// avoid frequent changes to mute & video layers
+	t.debouncer(t.UpdateVideoLayer)
+}
+
+func (t *SubscribedTrack) UpdateVideoLayer() {
+	t.updateDownTrackMute()
+	if t.subMuted.Get() || t.dt.Kind() != webrtc.RTPCodecTypeVideo {
+		return
+	}
+	settings, ok := t.settings.Load().(*livekit.UpdateTrackSettings)
+	if !ok {
+		return
+	}
+
+	quality := settings.Quality
+	if settings.Width > 0 {
+		quality = t.publishedTrack.GetQualityForDimension(settings.Width, settings.Height)
+	}
+	t.dt.SetMaxSpatialLayer(spatialLayerForQuality(quality))
```

Comment:
I am thinking at some point the full resolution + track source needs to be passed in so that stream allocator can use those to prioritize. Gut feel: just the one enum signal may not be enough.

Reply:
sounds good. we can change the interface to pass in add. details.